File: Divide in Incremental Groups[27-March-2023]/c++/Memoization.cpp
//		>> C++ Memoization Code <<

vector<vector<vector<int>>> dp;
    
    int solve(int i, vector<int>& arr, int N, int K, int sum) {
        if (sum > N) {
            return 0;
        }
        if (K == 0) {
            if (sum == N) {
                return dp[i][sum][K] = 1;
            } else {
                return dp[i][sum][K] = 0;
            }
        }
        if (i == arr.size()) {
            return dp[i][sum][K] = 0;
        }
        
        if (dp[i][sum][K] != -1) {
            return dp[i][sum][K];
        }
        
        return dp[i][sum][K] = solve(i, arr, N, K-1, sum+arr[i]) + solve(i+1, arr, N, K, sum);
    }
    
    int countWaystoDivide(int N, int K) {
        dp.resize(N+1, vector<vector<int>>(N+1, vector<int>(K+1, -1)));
        vector<int> arr(N);
        for (int i = 0; i < N; i++) {
            arr[i] = i+1;
        }
        return solve(0, arr, N, K, 0);
    }

File: Divide in Incremental Groups[27-March-2023]/java/Memoization.java

//        >> Memoization Java Code <<

class Solution {
    Integer[][][] dp;
    public int solve(int i,int[] arr,int N,int K,int sum)
    {
        if(sum>N) return 0;
        if(K==0)
        {
            if(sum==N)
                return dp[i][sum][K]=1;
            else
                return dp[i][sum][K]=0;
        }
        if(i==arr.length)
        {
            return dp[i][sum][K]=0;
        }
        
        if(dp[i][sum][K]!=null) return dp[i][sum][K];
        
        return dp[i][sum][K]=solve(i,arr,N,K-1,sum+arr[i])+solve(i+1,arr,N,K,sum);
        
    }
    public int countWaystoDivide(int N, int K) {
        
        dp=new Integer[N+1][N+1][K+1];
        int[] arr=new int[N];
        
        for(int i=0;i<N;i++) arr[i]=i+1;
        
        return solve(0,arr,N,K,0);
        
    }
}